internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -155,38 +179,156 @@ func (f *fleetGateway) Run(ctx context.Context) error {
 			return ctx.Err()
 		case <-f.scheduler.WaitTick():
 			f.log.Debug("FleetGateway calling Checkin API")
+			func() {
+				// get current state
+				stateUpdateSubCtx, cancelStateUpdateSub := context.WithCancel(ctx)
+				stateUpdateCh := f.stateFetcher.StateSubscribe(stateUpdateSubCtx).Ch()
+				defer func() {
+					cancelStateUpdateSub()
+				}()
+				state := <-stateUpdateCh
+
+				var checkinResponse *fleetapi.CheckinResponse
+
+				for checkinResponse == nil {
+					if ctx.Err() != nil {
+						return
+					}
+
+					checkinResult := f.performCancellableCheckin(ctx, state, stateUpdateCh, backoff, f.debouncerFactory())
+
+					if checkinResult.err != nil {
+						var relaunchCheckinErr *needNewCheckinError
+						if errors.As(checkinResult.err, &relaunchCheckinErr) {
+							// checkin was cancelled because we have an updated state to send to fleet
+							state = relaunchCheckinErr.newState
+						}
+
+						continue
+					}
+
+					// if we got here the checkin has been completed
+					checkinResponse = checkinResult.response
+				}
+
+				f.log.Debugf("Processing checkin response %v", checkinResponse)
+				actions := make([]fleetapi.Action, len(checkinResponse.Actions))
+				copy(actions, checkinResponse.Actions)
+				if len(actions) > 0 {
+					f.actionCh <- actions
+				}
+			}()
```

Comment:
Did you forget the `go`? otherwise why are you defining and immediately after invoking a anonymous function?
If it to use defer and return?

Reply:
To make this cleaner and easier to read it would be better to just move this to its own function.

Reply:
Done